File: src/net/dinkla/raytracer/materials/GlossyReflector.java
package net.dinkla.raytracer.materials;

import net.dinkla.raytracer.brdf.BRDF;
import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.colors.RGBColor;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.brdf.GlossySpecular;
import net.dinkla.raytracer.math.Ray;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.samplers.Sampler;
import net.dinkla.raytracer.worlds.World;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 25.04.2010
 * Time: 17:14:03
 * To change this template use File | Settings | File Templates.
 */
public class GlossyReflector extends Phong {

    protected final GlossySpecular glossySpecularBrdf;

    public GlossyReflector() {
        super();
        glossySpecularBrdf = new GlossySpecular();
    }

    public void setKr(final float kr) {
        glossySpecularBrdf.ks = kr;
    }

    public void setExp(final float exp) {
        glossySpecularBrdf.exp = exp;
    }

    @Override
    public Color areaLightShade(World world, Shade sr) {
        Color L = super.areaLightShade(world, sr);
        final Vector3D wo = sr.ray.getD().negate();
        final BRDF.Sample result = glossySpecularBrdf.sampleF(sr, wo);
        final Ray reflectedRay = new Ray(sr.getHitPoint(), result.wi);
        final Color r = world.getTracer().trace(reflectedRay, sr.depth +1);
        final Color r2 = result.color.mult(r);
        final Color r3 = r2.mult(result.wi.dot(sr.getNormal()) / result.pdf);
        return result.color;
//        L = L.plus(r);
//        return L;
    }

    public void setSampler(Sampler sampler) {
        glossySpecularBrdf.sampler = sampler;
    }
}

File: src/main/kotlin/net/dinkla/raytracer/materials/GlossyReflector.kt
package net.dinkla.raytracer.materials

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.brdf.GlossySpecular
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.samplers.Sampler
import net.dinkla.raytracer.world.World

class GlossyReflector : Phong() {

    protected val glossySpecularBrdf: GlossySpecular = GlossySpecular()

    // TODO kr == ks? it was like that in the orig code, check the book!
    var kr: Double
        get() = glossySpecularBrdf.ks
        set(v: Double) {
            glossySpecularBrdf.ks = v
        }

    override var exp : Double
        get() = glossySpecularBrdf.exp
        set(v: Double) {
            glossySpecularBrdf.exp = v
        }

    override fun areaLightShade(world: World, sr: Shade): Color {
        val L = super.areaLightShade(world, sr)
        val wo = -sr.ray.direction
        val result = glossySpecularBrdf.sampleF(sr, wo)
        val reflectedRay = Ray(sr.hitPoint, result.wi)
        val r = world.tracer.trace(reflectedRay, sr.depth + 1)
        val r2 = result.color * r
        val r3 = r2.times((result.wi dot sr.normal) / result.pdf)
        val c = result.color
        return c ?: Color.BLACK
        //        L = L.plus(r);
        //        return L;
    }

}
